examples/MetropolisWithinGibbs.md

```diff
@@ -0,0 +1,208 @@
+    [0., 1., 0., .8],
+    [.8, 0., 1., 0.],
+    [0., .8, 0., 1.]
+])
+
+def logprob(x, y):
+    return logprob_fn(x, y, Sigma=Sigma)
+```
+
+## MWG Sampling in BlackJAX
+
+In this case the conditional distributions $p(\xx \mid \yy)$ and $p(\yy \mid \xx)$ can be drawn from analytically (they are normal distributions).  However, for illustrative purposes we'll use an MCMC kernel to draw from each.  Specifically, we'll use an RMH kernel to draw from $p(\xx \mid \yy)$ and an HMC kernel to draw from $p(\yy \mid \xx)$.  To implement the corresponding MWG algorithm in BlackJAX, we'll write an `mwg_kernel()` for the problem which will do the following:
+
+1.  Maintain separate MCMC kernels to update each component of $p(\xx, \yy)$ while holding the other fixed.
+2.  Apply the kernel updates correctly.
+
+The issue with (2) is that each kernel update for a given MCMC `Algorithm` in BlackJAX refers to an algorithm-specific `AlgorithmState`.  For example, `RMHState` is a `typing.NamedTuple` class containing elements `position` and `log_probability`.  In our MWG sampling problem at the beginning of step $t$, `RMHState.log_probability` will consist of $\log p(\xx_{t-1}, \yy_{t-1})$.  After updating $\xx$, it will consist of $\log p(\xx_{t}, \yy_{t-1})$.  This happens automatically when we call `blackjax.mcmc.rmh.kernel()`.  However, after updating $\yy$ (via HMC), we must manually update `RMHState.log_probability` to consist of $\log p(\xx_{t}, \yy_{t})$.
+
+Two possible ways of doing this manual update are:
+
+1. Use the `blackjax.mcmc.algorithm.init()` function of the given component's MCMC algorithm to update the `AlgorithmState`.  This function has arguments `position` and `logprob_fn`.  In the case of the HMC component after obtaining $\xx_t$ but before drawing $\yy_t$, the `position` would be $\yy_{t-1}$ and the `logprob_fn` function would be $\log p(\xx_t, \cdot )$.
+
+2. The method above should work out-of-the-box for most (if not all) MCMC algorithms in BlackJAX.  However, it is not optimally efficient.  For example for the RMH update, after obtaining $\yy_{t-1}$ but before drawing $\xx_t$, the method above would calculate `RMHState.log_probability` to be $\log p(\xx_{t-1}, \yy_{t-1})$.  But we've already calculated this value from the previous HMC update of $\yy_{t-1} \sim p(\yy \mid \xx_{t-1})$.  So, we can save ourselves the cost of calculating the log-probability twice, at the expense of a deeper understanding of the low-level components of the algorithms at hand.
+
+We'll show how to implement each of these updates on one of the two MCMC kernels in the code below.
+
+### Construct the MWG Kernel
+
+```{code-cell} ipython3
+from blackjax.mcmc.hmc import HMCState
+from blackjax.mcmc.rmh import RMHState
+
+# MCMC updaters for each algorithm
+rmh_step_fn = blackjax.mcmc.rmh.kernel()
+hmc_step_fn = blackjax.mcmc.hmc.kernel()  # default integrator, etc.
+
+
+def mwg_kernel(rng_key, state_x, state_y, parameters_x, parameters_y):
```

Comment:
From an interface perspective, I think it would make sense to pass one big `state` and parameters individually, so that this looks like the other lower-level kernels in blackjax:

```python
def mwg_kernel(rng_key, state, sigma, step_size, inverse_mass_matrix, num_integration_steps)
```

where

```python
state = {'rmh': ..., 'hmc':}
```

what do you think?

Reply:
Makes sense for `state`.  For parameters, if we have two `rmh` components we would need to have arguments`sigma_x` and `sigma_y`.  But I guess that's not a great reason to break form with the lower-level BlackJAX kernels.

Perhaps a better reason to have `state` and `parameters` both of the form 
```python
state = {'x': ..., 'y': ...}
parameters = {'x': dict_x, 'y': dict_y}
```
is that with more than two sub-kernels in the Gibbs sampler, this would make it easier to cycle through them in a loop, as opposed to manually copy-pasting as I did in the two-kernel example.  Then again, doing this here would be needlessly complicated for people referring to the "how to" for a simple example to build from.  So maybe this can be flagged for later, in case it is decided to add built-in Gibbs sampling functionality to BlackJAX. 

OK I'll go ahead and make the changes.

Reply:
Makes sense! With that you're just one step away from a general M-w-G implementation 😄